### PR TITLE
initial checkin

### PR DESCRIPTION
**Pull request by Andyli1007**

### CODE REVIEW

File: src/collectives/device/msccl_kernel_impl.h [hunk 1]
```diff
@@ -392,6 +398,9 @@ __device__ __forceinline__ void mscclRunInterpreter(
           return;
       }
       if (t->hasDependence && tid == nthreads-1){
+        if (mscclShmem.work.needsFence) {
```
Comment (yzygitzh) [inline src/collectives/device/msccl_kernel_impl.h]: Not required for AMD

File: src/collectives/msccl.cc [hunk 2]
```diff
@@ -48,18 +48,23 @@ ncclResult_t mscclRunAlgo(
...
+
   NCCLCHECK(mscclSetupCount(hostAlgo, comm, count, dataType));
 
   NCCLCHECK(mscclSetupScratch(hostAlgo, stream));
 
   NCCLCHECK(mscclSetupSyncFlags(stream));
 
   if (status.connectedAlgos[comm].find(mscclAlgoHandle) == status.connectedAlgos[comm].end()) {
+    cudaStreamCaptureMode mode = cudaStreamCaptureModeRelaxed;
+    CUDACHECK(cudaThreadExchangeStreamCaptureMode(&mode));
```
Comment (yzygitzh) [inline src/collectives/msccl.cc]: Rename to hip

File: src/misc/msccl/msccl_setup.cc [hunk 3]
```diff
@@ -147,6 +180,39 @@ ncclResult_t mscclSetupProxy(struct mscclAlgo* hostAlgo, ncclComm_t comm) {
...
+      INFO(NCCL_INIT|NCCL_NET,"mscclSetupProxy: adding callback\n");
+
+      hipGraphNode_t callbackNode;
+      hipHostNodeParams p;
+      p.fn = mscclSetupProxyCallback;
+      auto params = &savedProxyArgs[threadLocalStatus.captureId];
+      p.userData = params;
+      CUDACHECK(hipGraphAddHostNode(&callbackNode, threadLocalStatus.graph, nullptr, 0, &p));
+    }
+    bool justInquire = false;
```
Comment (yzygitzh) [inline src/misc/msccl/msccl_setup.cc]: Need merge https://github.com/Azure/msccl-executor-nccl/commit/0f9e81727eb9d318c877f0697c1e5052dc73c033 as well

Comment (yzygitzh) [reply src/misc/msccl/msccl_setup.cc]: PR link: https://github.com/Azure/msccl-executor-nccl/pull/18/files

File: src/collectives/msccl.cc (hunk 2 above)
Comment (Andyli1007) [reply src/collectives/msccl.cc]: replaced

File: src/misc/msccl/msccl_setup.cc (hunk 3 above)
Comment (Andyli1007) [reply src/misc/msccl/msccl_setup.cc]: updated

**Review by yzygitzh — APPROVED** [review]

**Pull request merged**